src/Native/System.Net.Security.Native/pal_ntlmapi.cpp

```diff
@@ -0,0 +1,267 @@
+static int32_t
+NetSecurity_build_ntlm2_master(uint8_t* key, int32_t keylen, ntlm_buf* blob, ntlm_buf* sessionKey, ntlm_buf* masterKey)
+{
+    // reference: https://msdn.microsoft.com/en-us/library/cc236709.aspx
+    uint8_t* ntlmv2hash = NetSecurity_HMACDigest(key, keylen, blob->data, blob->length);
+    int32_t status = heim_ntlm_build_ntlm1_master(ntlmv2hash, UnsignedCast(keylen), sessionKey, masterKey);
+    if (status)
+    {
+        delete[] ntlmv2hash;
+        return status;
+    }
+
+    uint8_t* exportKey = NetSecurity_EVPEncrypt(ntlmv2hash, sessionKey->data, sessionKey->length);
+    delete[] ntlmv2hash;
+    masterKey->length = sessionKey->length;
+    masterKey->data = exportKey;
+    return status;
+}
+
+extern "C" int32_t NetSecurity_CreateType3Message(ntlm_buf* key,
+                                                  ntlm_type2* type2,
+                                                  const char* username,
+                                                  const char* domain,
+                                                  uint32_t flags,
+                                                  ntlm_buf* lmResponse,
+                                                  ntlm_buf* ntlmResponse,
+                                                  uint8_t* baseSessionKey,
+                                                  int32_t baseSessionKeyLen,
+                                                  ntlm_buf** outSessionHandle,
+                                                  int* outSessionKeyLen,
+                                                  ntlm_buf** outBufferHandle,
+                                                  int32_t* outLength)
+{
+    assert(key != nullptr);
+    assert(type2 != nullptr);
+    assert(lmResponse != nullptr);
+    assert(ntlmResponse != nullptr);
+    static char* workstation = static_cast<char*>(calloc(1, sizeof(char))); // empty string
+    ntlm_type3 type3;
+    memset(&type3, 0, sizeof(ntlm_type3));
+    type3.username = const_cast<char*>(username);
+    type3.targetname = const_cast<char*>(domain);
+    type3.lm = *lmResponse;
+    type3.ntlm = *ntlmResponse;
+    type3.ws = workstation;
+    type3.flags = flags;
+
+    int32_t status = 0;
+    ntlm_buf masterKey = {.length = 0, .data = nullptr};
+    *outSessionHandle = new ntlm_buf();
+
+    if (type2->targetinfo.length == 0)
+    {
+        status = heim_ntlm_build_ntlm1_master(key->data, key->length, *outSessionHandle, &masterKey);
+    }
+    else
+    {
+        // Only first 16 bytes of the NTLMv2 response should be passed
+        assert(ntlmResponse->length >= MD5_DIGEST_LENGTH);
+        ntlm_buf blob = {.length = MD5_DIGEST_LENGTH, .data = ntlmResponse->data};
+        status =
+            NetSecurity_build_ntlm2_master(baseSessionKey, baseSessionKeyLen, &blob, *outSessionHandle, &masterKey);
+    }
+    status = NetSecurity_SetBufferLength(status, outSessionHandle, outSessionKeyLen);
+    if (status != 0)
+    {
+        return status;
```

Comment:
Should outSessionHandle be freed here, and the output value set to nullptr?  (aka why is this if different than the next one).